// dimensions/packets/dumbpacketwriter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const bufferwriter_1 = require("dimensions/packets/bufferwriter");
class PacketWriter extends bufferwriter_1.default {
    constructor(size) {
        super(size);
    }
    updateSize() {
        const offset = this._offset;
        this._offset = 0;
        super.packUInt16(offset);
        this._offset = offset;
    }
    setType(type) {
        this.packUInt16(3);
        this.packByte(type);
        return this;
    }
    packInt16(int16) {
        super.packInt16(int16);
        this.updateSize();
        return this;
    }
    packUInt16(uint16) {
        super.packUInt16(uint16);
        this.updateSize();
        return this;
    }
    packInt32(int32) {
        super.packInt32(int32);
        this.updateSize();
        return this;
    }
    packUInt32(uint32) {
        super.packUInt32(uint32);
        this.updateSize();
        return this;
    }
    packSingle(single) {
        super.packSingle(single);
        this.updateSize();
        return this;
    }
    packByte(byte) {
        super.packByte(byte);
        this.updateSize();
        return this;
    }
    packSByte(byte) {
        super.packSByte(byte);
        this.updateSize();
        return this;
    }
    packHex(hex) {
        super.packHex(hex);
        this.updateSize();
        return this;
    }
    packBuffer(buffer) {
        super.packBuffer(buffer);
        this.updateSize();
        return this;
    }
    packString(str) {
        super.packString(str);
        this.updateSize();
        return this;
    }
}
exports.default = PacketWriter;
